File: 김호진/Week2-1/계단 오르기.py
# 백준 2579 계단 오르기 실3

import sys
input = sys.stdin.readline

n = int(input())
arr = list(int(input()) for _ in range(n))

dp = [0] * n
dp[0] = arr[0]
if n >= 2:
    dp[1] = arr[1] + arr[0]
if n >= 3:
    dp[2] = arr[2] + max(arr[1], arr[0])
if n >= 4:
    for i in range(3, n):
        dp[i] = arr[i] + max(arr[i-1] + dp[i-3], dp[i-2])

print(dp[n-1])

File: 한지웅/240902/계단오르기.js
// const input = require('fs').readFileSync("../input.txt").toString();
const input = require('fs').readFileSync("/dev/stdin").toString();

// data 첫번째는 계단의 step 수, 그 이후부터는 점수를 의미합니다.
const data = input.split('\n').map(Number);
const steps = data[0];
const scores = data.slice(1, data.length);
scores.unshift(0);

// dfs => 시간초과, DP활용
const dp = new Array(steps + 1).fill(0);

dp[1] = scores[1];
dp[2] = Math.max(scores[1] + scores[2], scores[2]);
dp[3] = Math.max(scores[1] + scores[3], scores[2] + scores[3]);
for (let j = 4; j < data.length; j++) {
    dp[j] = Math.max(scores[j-1] + dp[j-3], dp[j-2])
    dp[j] = dp[j] + scores[j]
}
console.log(dp[steps])
